firedrake/mesh.py

```diff
@@ -2415,6 +2415,37 @@ def clear_cell_sizes(self):
         except AttributeError:
             pass
 
+    @utils.cached_property
+    def cell_volume(self):
+        """
+        A :class:`~.Function` in the :math:`P^0` space containing the local mesh volume.
+
+        This is computed by interpolating the UFL :class:`~.CellVolume` for the current
+        mesh.
+        """
+        from firedrake.function import Function
+        from firedrake.functionspace import FunctionSpace
+        DG0 = FunctionSpace(self, "Discontinuous Lagrange", 0)
+        volume = Function(DG0)
+        return volume.interpolate(ufl.CellVolume(self))
```

Comment:
Is this actually mathematically correct for higher order meshes? Shouldn't it be a projection?

Reply:
As long as `CellVolume` is correctly integrated then this should be fine. Recall that `CellVolume` is intrinsically a DG0 function on any mesh.